Add trax sound inclusion as an argument.

Sasbom opened the pull request:
First of all: This is a great tool to have!
I'm very happy that it's still being maintained.

That aside, I found myself missing access to use the trax audio. Several projects I have worked on use the trax editor to throw together audio for playblasting, since it doesn't have to be sourced back to one sound node, and can therefore be layered.
Hence, why I added it, along with documentation within the docstring.

Have a good day,

Sas

Sasbom committed:
Add trax sound inclusion as an argument.

mottosso commented:
Thanks, could we not include this flag when passing `sound=True`? Or is there a usecase for having sound=True, trax_sounds=False or vice versa?

BigRoy commented:
> Thanks, could we not include this flag when passing `sound=True`? Or is there a usecase for having sound=True, trax_sounds=False or vice versa?

I believe sound could also come from the timeline as a single audio file, which as far as I know is separate from what audio is from `trax_sounds`. But not entirely sure.

Sasbom commented:
> Thanks, could we not include this flag when passing `sound=True`? Or is there a usecase for having sound=True, trax_sounds=False or vice versa?

So, i played around a little, it seems that when an audio file gets added to the trax editor, it gets imported as an audio clip too.
When you add an audio clip to the scene, it also appears in the trax editor.
The first track I imported through `Trax Editor > File > Import Audio`
and for the second track, I just dragged in into Maya.
![image](https://github.com/user-attachments/assets/7fcf2db4-3846-404b-92c6-98a6fd7c4f27)

I thought about typechecking the sound argument to switch the kwargs dict internally to use the useTraxSounds argument, but i opted to go for a seperate argument instead, because I noticed that the rest of the `capture` function follows maya's `cmds.playblast` arguments. I think it would be confusing if it didn't follow that.

You can't run useTraxSounds in combination with the sound argument either way.
I was under the impression that a sound node could be used in addition to the TRAX sound, for purposes of temporarily turning a track on during playblasting, however it seems that maya has made these arguments exclusive to eachother.
```
# Warning:  Use TRAX sound is set. The sound flag will be ignored.
```
Of course, we could just do something along the lines of:
```py
if isinstance(sound, bool):
  kwargs["useTraxSounds"] = sound
else:
  kwargs["sound"] = sound
```
or
```py
if sound is not None:
  snd_key = "uts" if isinstance(sound, bool) else "s"
  kwargs[snd_key] = sound
```

Since the arguments are mutually exclusive I'll push a commit to my fork to reflect the "sound=True" use.

Sasbom committed:
Used simple `isinstance` typecheck to confirm bool being passed to sound
argument, instead of creating a seperate trax_sound argument, since
these are exclusive in use in the native maya.cmds.playblast.

Sasbom committed:
Cleared up sound argument documentation.

mottosso commented:
It looks like the code got auto-formatted by something; there are far to many changes unrelated to the addition for me to merge this. If you can find a way to add the feature, without changing surrounding code, it would be easier to consider.

Sasbom committed:
revert accidental formatting.

Sasbom committed:
Merge branch 'abstractfactory:master' into feature/trax-sound

Sasbom commented:
> It looks like the code got auto-formatted by something; there are far to many changes unrelated to the addition for me to merge this. If you can find a way to add the feature, without changing surrounding code, it would be easier to consider.

Hello! Thank you for updating me. I synced my code with the main branch (for yesterdays' included PR), undid the formatting, the diff should look much more reasonable. I worked on this on another PC of mine that has autoformatting globally enabled, instead of a per project config. Rookie mistake.

mottosso commented:
Thank you, yes, much more readable now. @BigRoy, feel free to merge when happy!

BigRoy commented on capture.py:
```diff
@@ -72,8 +72,13 @@ def capture(camera=None,
-            playblast. When None (default) no sound will be used.
+        sound (str, bool, optional):  Specify the sound node to be used
+            during playblast.
+            When None (default) no sound node will be used,
+            except for sound that is linked in any present Time Editor Composition
+            Audio clip.
```
What's this `Time Editor Composition Audio clip`, and why is it still present when `None`? (Doesn't seem like a change in functionality. I'm just not sure whether it's entirely clear what this added comment refers to.)

Sasbom replied:
Hi,
this is essentially referring to what you were talking about with

> I believe sound could also come from the timeline...

Once a sound file is added to the timeline directly, maya creates an audio track in a composition structure containing the file itself, which is distinct from the audio nodes. 
```
timeEditor
- Compositions
  - Composition#
     - audio_track#
       - file_[file_name]
```
I tested out the behavior of this with the regular playblast command, and it seems that maya includes whatever is in here by default. 
I added the comment to clarify that users could expect audio to show up in their file if it was set in this way, because that's the default behavior of the playblast command.

It'd be neat to look at whether these audiotracks are queryable, to maybe manage them with a context manager, but so far I haven't quite figured out how to access these.

I should probably change `clip` to `track`, since that is the "official" naming scheme.

BigRoy replied:
Thanks, that does clarify things!

> since that is the "official" naming scheme.

Any chance you have some links to this official naming/documentation? Or where are you getting that structure from? Is that visible somewhere inside Maya?

Anyway, since it's not new behavior but only elaborating on what it would be doing it's a good addition, so thanks! I just wish the Maya `playblast` docs would've elaborated on this too so we're just aligning to their docs.

Anyway - for reference, here's [a recent `playblast` docs](https://help.autodesk.com/cloudhelp/ENU/MayaCRE-Tech-Docs/CommandsPython/playblast.html)

This flag did stand out to me: [`combineSound`](https://help.autodesk.com/cloudhelp/ENU/MayaCRE-Tech-Docs/CommandsPython/playblast.html#flagcombineSound):
> Combine the trax sounds into a single track. This might force a resampling of the sound if the sound paramters don't match up. 

Is that relevant to use with this PR too?

Sasbom replied:
> Any chance you have some links to this official naming/documentation? Or where are you getting that structure from? Is that visible somewhere inside Maya?

I'm not sure i can find it in the official maya documentation, but it does reflect native behavior as shown here:
![image](https://github.com/user-attachments/assets/08fda887-187b-4fa4-8113-6b4ffc62fa7a)
This appeared in the DAG viewer upon creating an audio track in the Time Editor.
[The documentation](https://help.autodesk.com/view/MAYAUL/2025/ENU/?guid=GUID-E4B5DB7D-7351-4561-BD8B-60AC9D48DDF6) isn't very clear about this.

> This flag did stand out to me: [`combineSound`](https://help.autodesk.com/cloudhelp/ENU/MayaCRE-Tech-Docs/CommandsPython/playblast.html#flagcombineSound):
> 
> > Combine the trax sounds into a single track. This might force a resampling of the sound if the sound paramters don't match up.
> 
> Is that relevant to use with this PR too?

I looked at this and did some research to see if I can find any practical application for this, and I really can't. I often see the Trax editor being used as a sound source on projects that have multiple audio nodes arranged together, and playblasting with or without this flag doesn't change much except
> This might force a resampling of the sound if the sound paramters don't match up.
which, spelling mistake aside, I don't think I've ever deemed nessecary.
I tried to search if anyone since the flags existence since maya 2011 has ever needed it to solve a problem, and I can not find anything on google, only links to the multitudes of maya documentations hosted online.

Therefore, I think it's irrelevant to this. 
If it turns out to be relevant it might be interesting to make a split in the sound and trax_sound argument again, since right now, we're providing passing a boolean value as a shortcut.

BigRoy replied:
Thanks! 🕵️

Then only what remains is, what is the _exact_ reason why we're deviating from the arguments that `maya.cmds.playblast` has - why wouldn't we just add the `useTraxSounds` argument and call it a day and have anyone wanting to use these features just rely on their knowledge of `maya.cmds.playblast`? 

@mottosso what's your thought? It may simplify the usage of `capture` yet at the same time also means you need to learn more about the argument of `capture` itself when compared to e.g. knowledge of `maya.cmds.playblast` of a TD.

mottosso replied:
> @mottosso what's your thought?

Hm, I can't recall whether we were trying to be a drop-in replacement for `cmds.playblast`? Do most argument already line up? I'd imagine it's not relevant, although I can see the use.

I'm happy to leave the decision to you Roy, best I can do is give an opinion but would not suffer the consequences of it like you would.

Sasbom committed:
capture.py sound argument description update to more accurately reflect reference to "audio track" instances in maya's Time Editor

BigRoy reviewed (approved):
Let's do it as it is. Thanks!
